torch/_dynamo/graph_deduplication.py

```diff
@@ -0,0 +1,201 @@
+for each output, and all nodes in the region with external outputs are replaced by the proper \
+getitem node. Finally, all original nodes are erased (there should be no uses of these \
+left in the graph).
+
+The deduplication mutates the output_graph argument in place.
+
+Returns a mapping of nodes to their subgraph output replacement node to remap outputs
+when they are created in output_graph.
+    """
+    duplicated_region_groups = output_graph.region_tracker.get_identical_regions(
+        output_graph.graph
+    )
+
+    # Used to track which nodes were replaced with subgraph outputs
+    # today, we have to register the new subgraph submodules before the
+    # graph outputs have been created, so we pass the replacement mapping
+    # back to output graph to do the replacements at the site of output creation
+    output_replacements: Dict[Node, Node] = {}
+    for region_group in duplicated_region_groups:
+        inds_with_external_users = _get_all_output_indices(region_group)
+        region = region_group[0]
+        (
+            subgraph,
+            node_ind_arg_inds,
+        ) = _create_subgraph(region, inds_with_external_users)
+        sub_gm = torch.fx.GraphModule(output_graph.nn_modules, subgraph)
+        subgraph_name = output_graph.install_subgraph("subgraph", sub_gm)
+        with output_graph.graph.inserting_before():
+            get_subgraph_node = output_graph.graph.create_node(
+                "get_attr", subgraph_name, (), {}
+            )
+        for region in region_group:
+            try:
+                _replace_region_with_subgraph(
+                    output_graph.graph,
+                    region,
+                    get_subgraph_node,
+                    node_ind_arg_inds.keys(),
+                    inds_with_external_users,
+                    sub_gm,
+                    subgraph_name,
+                    output_replacements,
+                )
+            except Unsupported:
+                log.debug(
+                    "Failed to substitute region %s due to input alias or mutation",
+                    region,
+                )
+
+    return output_replacements
+
+
+def _replace_region_with_subgraph(
+    graph: torch.fx.Graph,
+    region: Region,
+    get_subgraph_node: Node,
+    node_ind_arg_ind: Iterable[Tuple[int, int]],
+    inds_with_external_users: List[int],
+    sub_gm: torch.fx.GraphModule,
+    subgraph_name: str,
+    output_replacements: Dict[Node, Node],
+) -> None:
+    sub_args = []
+    for node_ind, arg_ind in node_ind_arg_ind:
+        node = region[node_ind]
+        flattened_args_kwargs, _ = tree_flatten((node.args, node.kwargs))
+        sub_args.append(flattened_args_kwargs[arg_ind])
+
+    invoke_args = (get_subgraph_node, subgraph_name, tuple(sub_args))
+    fake_inputs = [node.meta["example_value"] for node in sub_args]
+
+    if has_potential_input_alias_or_mutation(sub_gm, fake_inputs):
```

Comment:
Do you  have a test that a region with input alias / mutation correctly does not get put into an invoke_subgraph?

Reply:
No, I'll add